File: apps/RelayiumKit/Sources/RelayiumLocalPeerKit/LocalPeerTransportLifecycle.swift
import Foundation

/// Queue-confined decision state for a listener/browser pair.
struct LocalPeerTransportLifecycle {
    enum Phase: Equatable { case idle, starting, running, failed, stopped }
    enum StartDecision: Equatable { case arm, ignore }
    enum AnnouncementDecision: Equatable { case announce, ignore }
    enum StopDecision: Equatable { case tearDown, ignore }

    private(set) var phase: Phase = .idle
    private var listenerReady = false
    private var browserReady = false

    mutating func start() -> StartDecision {
        guard phase == .idle else { return .ignore }
        phase = .starting
        return .arm
    }

    mutating func listenerBecameReady() -> AnnouncementDecision {
        guard phase == .starting else { return .ignore }
        listenerReady = true
        return announceWhenReady()
    }

    mutating func browserBecameReady() -> AnnouncementDecision {
        guard phase == .starting else { return .ignore }
        browserReady = true
        return announceWhenReady()
    }

    mutating func fail() -> AnnouncementDecision {
        guard phase != .failed, phase != .stopped else { return .ignore }
        phase = .failed
        return .announce
    }

    /// The arming window closed with neither half ready and nothing failed.
    ///
    /// `NWListener` and `NWBrowser` do not report a refused Local Network
    /// permission — or a link that is simply not up — as `.failed`. They sit in
    /// `.waiting`, which is neither of the two edges this type announces, so
    /// without a deadline the channel never opens, `LanDiscoveryModel` never
    /// leaves `connecting`, and the user is shown a search that cannot end.
    /// Failing is the honest answer AND the recoverable one: the model's bounded
    /// backoff reopens, so a permission granted a moment later is picked up.
    ///
    /// Only from `starting`. A pair that became ready, failed, or was stopped
    /// has already had its say, and a late timer must not overrule it.
    mutating func startDeadlineElapsed() -> AnnouncementDecision {
        guard phase == .starting else { return .ignore }
        phase = .failed
        return .announce
    }

    mutating func stop() -> StopDecision {
        guard phase != .stopped else { return .ignore }
        phase = .stopped
        return .tearDown
    }

    var isDeliveringEvents: Bool { phase == .starting || phase == .running }

    private mutating func announceWhenReady() -> AnnouncementDecision {
        guard listenerReady, browserReady else { return .ignore }
        phase = .running
        return .announce
    }
}

File: apps/android/app/src/main/kotlin/com/relayium/android/nearby/LocalPeerTransportLifecycle.kt
package com.relayium.android.nearby

/**
 * Owner-thread-confined decision state for one advertise/browse pair.
 *
 * Pure, so the one rule that is easy to get wrong — a late timer must never
 * overrule an edge that already happened — is directly testable instead of
 * being inferred from an emulator run. Port of
 * `LocalPeerTransportLifecycle.swift`.
 */
class LocalPeerTransportLifecycle {

    enum class Phase { IDLE, STARTING, RUNNING, FAILED, STOPPED }
    enum class StartDecision { ARM, IGNORE }
    enum class AnnouncementDecision { ANNOUNCE, IGNORE }
    enum class StopDecision { TEAR_DOWN, IGNORE }

    var phase: Phase = Phase.IDLE
        private set

    private var advertiseReady = false
    private var browseReady = false

    fun start(): StartDecision {
        if (phase != Phase.IDLE) return StartDecision.IGNORE
        phase = Phase.STARTING
        return StartDecision.ARM
    }

    fun advertiseBecameReady(): AnnouncementDecision {
        if (phase != Phase.STARTING) return AnnouncementDecision.IGNORE
        advertiseReady = true
        return announceWhenReady()
    }

    fun browseBecameReady(): AnnouncementDecision {
        if (phase != Phase.STARTING) return AnnouncementDecision.IGNORE
        browseReady = true
        return announceWhenReady()
    }

    fun fail(): AnnouncementDecision {
        if (phase == Phase.FAILED || phase == Phase.STOPPED) return AnnouncementDecision.IGNORE
        phase = Phase.FAILED
        return AnnouncementDecision.ANNOUNCE
    }

    /**
     * The arming window closed with neither half ready and nothing failed.
     *
     * `NsdManager` does not report "this link has no multicast" — and, on a
     * future target, will not report a refused local-network permission — as a
     * registration or discovery FAILURE. It simply never calls back, so without
     * a deadline the channel never opens and the user is shown a search that
     * cannot end. Failing is the honest answer AND the recoverable one: the
     * model's bounded backoff reopens, so a link that comes up a moment later
     * is picked up.
     *
     * Only from STARTING: a pair that became ready, failed or was stopped has
     * already had its say.
     */
    fun startDeadlineElapsed(): AnnouncementDecision {
        if (phase != Phase.STARTING) return AnnouncementDecision.IGNORE
        phase = Phase.FAILED
        return AnnouncementDecision.ANNOUNCE
    }

    fun stop(): StopDecision {
        if (phase == Phase.STOPPED) return StopDecision.IGNORE
        phase = Phase.STOPPED
        return StopDecision.TEAR_DOWN
    }

    /** Whether delegate events may still be delivered. A listener callback that
     *  fires after a stop is a callback about a room that is gone. */
    val isDeliveringEvents: Boolean
        get() = phase == Phase.STARTING || phase == Phase.RUNNING

    private fun announceWhenReady(): AnnouncementDecision {
        if (!advertiseReady || !browseReady) return AnnouncementDecision.IGNORE
        phase = Phase.RUNNING
        return AnnouncementDecision.ANNOUNCE
    }
}
